storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,16 +3579,326 @@ dberr_t fsp_sys_tablespace_validate()
+      emplace(uint64_t{page} << 16 | offset).second;
+  }
+
+  __attribute__((warn_unused_result))
+  bool find(uint32_t page, uint16_t offset) const
+  { return std::unordered_set<uint64_t>::find(uint64_t{page} << 16 | offset) != end(); }
+};
+
+/** Determine the inodes used by tables in the system tablespace.
+@param mtr    mini-transaction
+@param inodes index nodes found
+@return error code */
+static dberr_t fsp_table_inodes(mtr_t *mtr, inode_info *inodes)
+{
+  btr_pcur_t pcur;
+  ulint len;
+  const auto savepoint= mtr->get_savepoint();
+  dberr_t err= DB_SUCCESS;
+  dict_sys.freeze(SRW_LOCK_CALL);
+  std::unordered_set<uint32_t> root_pages;
+  for (const rec_t *rec= dict_startscan_system(
+         &pcur, mtr, dict_sys.sys_indexes);
+       rec; rec= dict_getnext_system_low(&pcur, mtr))
+  {
+    const byte *field= rec_get_nth_field_old(
+      rec, DICT_FLD__SYS_INDEXES__SPACE, &len);
+    if (len != 4)
+    {
+      err= DB_CORRUPTION;
+      break;
+    }
+    uint32_t space= mach_read_from_4(field);
+    if (space > 0) continue;
+
+    field= rec_get_nth_field_old(
+      rec, DICT_FLD__SYS_INDEXES__PAGE_NO, &len);
+    if (len != 4)
+    {
+      err= DB_CORRUPTION;
+      break;
+    }
+    uint32_t root_page_no= mach_read_from_4(field);
+    if (root_page_no != FIL_NULL)
+      root_pages.insert(root_page_no);
+  }
+  mtr->rollback_to_savepoint(savepoint);
+  dict_sys.unfreeze();
+
+  if (err) return err;
+  /* Add system table index root pages */
+  dict_sys.sys_tables->add_index_root_pages(&root_pages);
+  dict_sys.sys_indexes->add_index_root_pages(&root_pages);
+  dict_sys.sys_columns->add_index_root_pages(&root_pages);
+  dict_sys.sys_fields->add_index_root_pages(&root_pages);
+  dict_sys.sys_foreign->add_index_root_pages(&root_pages);
+  dict_sys.sys_foreign_cols->add_index_root_pages(&root_pages);
+  dict_sys.sys_virtual->add_index_root_pages(&root_pages);
+
+ for (auto page : root_pages)
+ {
+   buf_block_t *block= buf_page_get_gen(
+     page_id_t{0, page}, 0, RW_NO_LATCH, nullptr, BUF_GET,
+     mtr, &err);
+   if (!block)
+     break;
+
+   if (!inodes->insert_seg(
+         block->page.frame + PAGE_HEADER + PAGE_BTR_SEG_TOP))
+   {
+     err= DB_CORRUPTION;
+     break;
+   }
+
+   if (!inodes->insert_seg(
+          block->page.frame + PAGE_HEADER + PAGE_BTR_SEG_LEAF))
+   {
+     err= DB_CORRUPTION;
+     break;
+   }
+ }
+ return err;
+}
+
+/* Get the used inode from the system tablespace
+@param inodes  inode information used found in system tablespace
+@return error code */
+dberr_t fsp_get_sys_used_segment(inode_info *inodes)
+{
+  mtr_t mtr;
+  dberr_t err= DB_SUCCESS;
+  buf_block_t *block= nullptr;
+  mtr.start();
+  /* Get TRX_SYS_FSEG_HEADER, TRX_SYS_DOUBLEWRITE_FSEG from
+  TRX_SYS_PAGE */
+  block= buf_page_get_gen(
+    page_id_t{0, TRX_SYS_PAGE_NO}, 0, RW_NO_LATCH, nullptr,
+    BUF_GET, &mtr, &err);
+  if (!block)
+  {
+err_exit:
+    mtr.commit();
+    return err;
+  }
```

Comment:
If we took `mtr` as a parameter, each `goto err_exit;` would reduce to `return err;`.

I think that it could be even better to use `buf_pool.page_fix()` to read the pages. I would like to get rid of the `RW_NO_LATCH` parameter of `buf_page_get_gen()` and use that function exclusively for latched block access. In that case we would have to remember to invoke `block->unfix()` after we are done with a block.